### PR TITLE
correct codegen bugs introduced by allocation-hoisting-PR

### PR DESCRIPTION
**Pull request by gbaraldi**
Adresses some of the reviews for #45153

### CODE REVIEW

**Comment by gbaraldi** [comment]
@vtjnash How does this look to you? The mangling from `emit_a_ccall` gave errors so I got it from some passes downstream.

File: src/cgutils.cpp [hunk 1]
```diff
@@ -901,7 +901,7 @@ static LoadInst *emit_nthptr_recast(jl_codectx_t &ctx, Value *v, ssize_t n, MDNo
         emit_bitcast(ctx, vptr, PointerType::get(type, 0)))));
  }
 
-static Value *boxed(jl_codectx_t &ctx, const jl_cgval_t &v);
+static Value *boxed(jl_codectx_t &ctx, const jl_cgval_t &v,  bool is_promotable=false);
```
Comment (vtjnash) [inline src/cgutils.cpp:904]: So promotion is now always forbidden implicitly?

File: src/cgutils.cpp [hunk 2]
```diff
@@ -3163,6 +3163,45 @@ static Value *box_union(jl_codectx_t &ctx, const jl_cgval_t &vinfo, const SmallB
...
+{
+    Intrinsic::ID ID = call->getIntrinsicID();
+    auto nargs = call->arg_size();
+    SmallVector<Value*, 8> args(nargs);
+    SmallVector<Type*, 8> argTys(nargs);
+    for (unsigned i = 0; i < nargs; i++) {
+        auto arg = call->getArgOperand(i);
+        args[i] = arg;
+        argTys[i] = args[i]->getType();
+    }
```
Comment (vtjnash) [inline src/cgutils.cpp:3176]: You should not be making an array of args if you don't need it
```suggestion
    SmallVector<Type*, 8> argTys(nargs);
    for (unsigned i = 0; i < nargs; i++) {
        auto argi = call->getArgOperand(i);
        argTys[i] = argi->getType();
    }
```

Comment (vtjnash) [reply src/cgutils.cpp:3176]: It is also slightly extraneous to iterate up to `nargs`, when you are going to drop everything after `oldfType->getNumParams()`

File: src/cgutils.cpp (hunk 1 above)
Comment (gbaraldi) [reply src/cgutils.cpp:904]: Yes, the regression we were seeing was from something getting promoted when we called `boxed` on a separate part fo the of the compilation and then hoist something that shouldn't be hoisted. 
In the case of the regression we were hosting a string allocation from inside an error to the main path of the function which then impeded an optimization.
From what I understand we should only hoist the allocation when we have nested structs and it's fields are promotable which we only check inside `emit_new_struct`

File: src/cgutils.cpp [hunk 3]
```diff
@@ -3639,13 +3672,14 @@ static jl_cgval_t emit_new_struct(jl_codectx_t &ctx, jl_value_t *ty, size_t narg
...
-                    } else if (!promotion_point) {
+                    }
+                    else if (!promotion_point) {
                         promotion_point = inst;
                     }
                 }
                 Value *fval = NULL;
                 if (jl_field_isptr(sty, i)) {
-                    fval = boxed(ctx, fval_info);
+                    fval = boxed(ctx, fval_info, is_promotable);
```
Comment (vtjnash) [inline src/cgutils.cpp:3682]:
```suggestion
                    fval = boxed(ctx, fval_info, field_promotable);
```
?

**Comment by gbaraldi** [comment]
@nanosoldier runbenchmarks(ALL, vs = ":master")

**Comment by fredrikekre** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/45476#issuecomment-1155077805):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(setenv(`git push`; dir="/nanosoldier/workdir/NanosoldierReports"), ProcessExited(1)) [1]

```
Unfortunately, the logs could not be uploaded.

**Comment by vtjnash** [comment]
https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c299527_vs_6c97a48/report.md

**Comment by gbaraldi** [comment]
@vtjnash I'm not super accustomed to nanosoldier results, was this a particularly noisy benchmark or should I take a look at some more regressions. 
The original regression found i.e more memory usage in skipmissing seems to have been fixed.

**Comment by vtjnash** [comment]
It looks like noise, though it wouldn't be too surprising if some of the differences were real. The memory changes are always real.

**Comment by inkydragon** [comment]
Maybe related to #45485.

```jl
julia> versioninfo()
Julia Version 1.9.0-DEV.883
Commit e4257cbc89 (2022-06-30 17:58 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 6 × Intel(R) Core(TM) i5-9400F CPU @ 2.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.5 (ORCJIT, skylake)
  Threads: 1 on 6 virtual cores

julia> VERSION
v"1.9.0-DEV.883"

julia> struct MaybeTuple
           val::Union{Nothing, Tuple{Float32}}
       end

julia> MaybeTuple((0,))
Assertion failed: New->getType() == getType() && "replaceAllUses of value with new value of different type!", file /workspace/srcdir/llvm-project/llvm/lib/IR/Value.cpp, line 503

signal (22): SIGABRT
in expression starting at REPL[4]:1
```

**Comment by gbaraldi** [comment]
Yeah, this branch currently breaks unions.

**Comment by vchuravy** [comment]
How close is this? Should we revert #45153 and then re-land it with these fixes?

**Comment by gbaraldi** [comment]
We might want to run pkgeval, but the issues related to the original pr are fixed by this

**Comment by vchuravy** [comment]
@nanosoldier `runtests(ALL, vs = ":master", buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"], vs_buildflags=["LLVM_ASSERTIONS=1", "FORCE_ASSERTIONS=1"])`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45476#issuecomment-1173150846) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2e8363b_vs_05eb153/report.html).

**Comment by gbaraldi** [comment]
After a rebase I tested it locally and the failures seen weren't there

File: src/cgutils.cpp [hunk 4]
```diff
@@ -3681,7 +3681,7 @@ static jl_cgval_t emit_new_struct(jl_codectx_t &ctx, jl_value_t *ty, size_t narg
                 if (fval_info.typ == jl_bottom_type)
                     return jl_cgval_t();
                 // TODO: Use (post-)domination instead.
-                bool field_promotable = !init_as_value && fval_info.promotion_ssa != -1 &&
+                bool field_promotable = !jl_is_uniontype(jtype) && !init_as_value && fval_info.promotion_ssa != -1 &&
```
Comment (vtjnash) [inline src/cgutils.cpp:3684]: It would be valid though if `jl_field_isptr`

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Review by vtjnash — APPROVED** [review]

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/45476#issuecomment-1178181796) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f8f7028_vs_cae7fef/report.html).

File: src/cgutils.cpp (hunk 4 above)
Comment (gbaraldi) [reply src/cgutils.cpp:3684]: The issue is that unions currently don't get their promotion point correctly, I'm not exactly what's needed to fix that, so for now, not allowing unions at all seems fine.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45476#issuecomment-1178181990) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/566c5f9_vs_cae7fef/report.md).

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(["array", "comprehension", ("collect", "Vector{Float64}")], vs=":master")`

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks("array" && "comprehension" && ("collect", "Vector{Float64}"), vs=":master")`

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks("array" && "comprehension", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45476#issuecomment-1178536805) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/a8a1dd4_vs_51bb968/report.md).

**Pull request merged**